src/kafka/producer/sender.cc

```diff
@@ -0,0 +1,261 @@
+                return std::make_pair(broker, response);
+        }));
+    }
+}
+
+void sender::set_error_code_for_broker(const sender::connection_id& broker, const error::kafka_error_code& error_code) {
+    for (auto& [topic, messages_by_partition] : _messages_split_by_broker_topic_partition[broker]) {
+        for (auto& [partition, messages] : messages_by_partition) {
+            for (auto& message : messages) {
+                message->_error_code = error_code;
+            }
+        }
+    }
+}
+
+void sender::set_error_code_for_topic_partition(const std::string& topic, int32_t partition_index,
+        const error::kafka_error_code& error_code) {
+    for (auto& message : _messages_split_by_topic_partition[{topic, partition_index}]) {
+        message->_error_code = error_code;
+    }
+}
+
+void sender::set_success_for_topic_partition(const std::string& topic, int32_t partition_index) {
+    for (auto& message : _messages_split_by_topic_partition[{topic, partition_index}]) {
+        message->_error_code = error::kafka_error_code::NONE;
+        message->_promise.set_value();
+    }
+}
+
+void sender::move_messages(std::vector<sender_message>& messages) {
+    _messages.insert(_messages.end(), std::make_move_iterator(messages.begin()),
+              std::make_move_iterator(messages.end()));
+    messages.clear();
+}
+
+size_t sender::messages_size() const {
+    return _messages.size();
+}
+
+bool sender::messages_empty() const {
+    return _messages.empty();
+}
+
+void sender::send_requests() {
+    split_messages();
+    queue_requests();
+}
+
+future<> sender::receive_responses() {
+    return when_all(_responses.begin(), _responses.end()).then([this](auto responses){
+        set_error_codes_for_responses(responses);
+        filter_messages();
+        return process_messages_errors();
+    });
+}
+
+future<> sender::process_messages_errors() {
+    auto should_refresh_metadata = false;
+    for (auto& message : _messages) {
+        if ((*message._error_code)._is_invalid_metadata) {
+            should_refresh_metadata = true;
+            break;
+        }
+    }
+    if (should_refresh_metadata) {
+        return _metadata_manager->refresh_metadata().discard_result();
+    } else {
+        return make_ready_future<>();
+    }
+}
+
+void sender::filter_messages() {
+    _messages.erase(std::remove_if(_messages.begin(), _messages.end(), [](auto& message) {
+        if (message._error_code == error::kafka_error_code::NONE) {
+            return true;
+        }
+        if (!(*message._error_code)._is_retriable) {
+            message._promise.set_exception(send_exception((*message._error_code)._error_message));
+            return true;
+        }
+        return false;
+    }), _messages.end());
+}
+
+void sender::set_error_codes_for_responses(std::vector<future<std::pair<connection_id, produce_response>>>& responses) {
+    for (auto& response : responses) {
+        auto [broker, response_message] = response.get0();
```

Comment:
Aren't `get0`s only working in a seastar thread environment?

Reply:
Yes, but `set_error_codes_for_responses` is invoked with output value of `when_all` (https://github.com/avelanarius/seastar/pull/14/files/ada9c5e22c0dc32242018b3489a3de57470d1539#diff-fc4f027ee3f5e8a489ca1fd1dcd1a1f4R198), so the futures have finished running (it's safe to `get0()`).